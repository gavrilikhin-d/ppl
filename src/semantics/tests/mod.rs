use crate::test_compilation_result;

test_compilation_result!(candidate_not_viable);
test_compilation_result!(constraints);
test_compilation_result!(constraints_in_constructor);
test_compilation_result!(generics);
test_compilation_result!(missing_fields);
test_compilation_result!(multiple_errors);
test_compilation_result!(multiple_initialization);
test_compilation_result!(non_class_constructor);
test_compilation_result!(predeclare_vars);
test_compilation_result!(references);
test_compilation_result!(traits);
test_compilation_result!(type_as_value);
test_compilation_result!(wrong_initializer_type);
