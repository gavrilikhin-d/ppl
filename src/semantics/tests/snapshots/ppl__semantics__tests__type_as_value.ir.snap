---
source: src/semantics/tests/mod.rs
expression: ir
---
; ModuleID = 'type_as_value'
source_filename = "type_as_value.ppl"

%"Type<Integer>" = type { %String, %Integer }
%String = type { ptr }
%Integer = type { ptr }
%"Type<None>" = type { %String, %Integer }
%"Type<Bool>" = type { %String, %Integer }
%"Type<String>" = type { %String, %Integer }
%"Type<Type<Integer>>" = type { %String, %Integer }

@"Type<Integer>" = global %"Type<Integer>" zeroinitializer
@0 = private unnamed_addr constant [8 x i8] c"Integer\00", align 1
@"Type<None>" = global %"Type<None>" zeroinitializer
@1 = private unnamed_addr constant [5 x i8] c"None\00", align 1
@"Type<Bool>" = global %"Type<Bool>" zeroinitializer
@2 = private unnamed_addr constant [5 x i8] c"Bool\00", align 1
@"Type<String>" = global %"Type<String>" zeroinitializer
@3 = private unnamed_addr constant [7 x i8] c"String\00", align 1
@"Type<Type<Integer>>" = global %"Type<Type<Integer>>" zeroinitializer
@4 = private unnamed_addr constant [14 x i8] c"Type<Integer>\00", align 1
@x = global %"Type<Integer>" zeroinitializer
@y = global %Integer zeroinitializer
@5 = private unnamed_addr constant [5 x i8] c"true\00", align 1
@6 = private unnamed_addr constant [6 x i8] c"false\00", align 1

define void @initialize() !dbg !3 {
  %1 = alloca %"Type<Integer>", align 8, !dbg !7
  %"Type<Integer>.name" = getelementptr inbounds %"Type<Integer>", ptr %1, i32 0, i32 0, !dbg !7
  %2 = call %String @string_from_c_string_and_length(ptr @0, i64 7), !dbg !8
  store %String %2, ptr %"Type<Integer>.name", align 8, !dbg !8
  %"Type<Integer>.size" = getelementptr inbounds %"Type<Integer>", ptr %1, i32 0, i32 1, !dbg !8
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !8
  store %Integer %3, ptr %"Type<Integer>.size", align 8, !dbg !8
  %4 = load %"Type<Integer>", ptr %1, align 8, !dbg !8
  store %"Type<Integer>" %4, ptr @"Type<Integer>", align 8, !dbg !8
  br label %return, !dbg !8

return:                                           ; preds = %0
  ret void
}

declare %String @string_from_c_string_and_length(ptr, i64)

declare %Integer @integer_from_i64(i64)

define void @initialize.1() !dbg !9 {
  %1 = alloca %"Type<None>", align 8, !dbg !10
  %"Type<None>.name" = getelementptr inbounds %"Type<None>", ptr %1, i32 0, i32 0, !dbg !10
  %2 = call %String @string_from_c_string_and_length(ptr @1, i64 4), !dbg !11
  store %String %2, ptr %"Type<None>.name", align 8, !dbg !11
  %"Type<None>.size" = getelementptr inbounds %"Type<None>", ptr %1, i32 0, i32 1, !dbg !11
  %3 = call %Integer @integer_from_i64(i64 0), !dbg !11
  store %Integer %3, ptr %"Type<None>.size", align 8, !dbg !11
  %4 = load %"Type<None>", ptr %1, align 8, !dbg !11
  store %"Type<None>" %4, ptr @"Type<None>", align 8, !dbg !11
  br label %return, !dbg !11

return:                                           ; preds = %0
  ret void
}

define void @initialize.2() !dbg !12 {
  %1 = alloca %"Type<Bool>", align 8, !dbg !13
  %"Type<Bool>.name" = getelementptr inbounds %"Type<Bool>", ptr %1, i32 0, i32 0, !dbg !13
  %2 = call %String @string_from_c_string_and_length(ptr @2, i64 4), !dbg !14
  store %String %2, ptr %"Type<Bool>.name", align 8, !dbg !14
  %"Type<Bool>.size" = getelementptr inbounds %"Type<Bool>", ptr %1, i32 0, i32 1, !dbg !14
  %3 = call %Integer @integer_from_i64(i64 1), !dbg !14
  store %Integer %3, ptr %"Type<Bool>.size", align 8, !dbg !14
  %4 = load %"Type<Bool>", ptr %1, align 8, !dbg !14
  store %"Type<Bool>" %4, ptr @"Type<Bool>", align 8, !dbg !14
  br label %return, !dbg !14

return:                                           ; preds = %0
  ret void
}

define void @initialize.3() !dbg !15 {
  %1 = alloca %"Type<String>", align 8, !dbg !16
  %"Type<String>.name" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 0, !dbg !16
  %2 = call %String @string_from_c_string_and_length(ptr @3, i64 6), !dbg !17
  store %String %2, ptr %"Type<String>.name", align 8, !dbg !17
  %"Type<String>.size" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 1, !dbg !17
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !17
  store %Integer %3, ptr %"Type<String>.size", align 8, !dbg !17
  %4 = load %"Type<String>", ptr %1, align 8, !dbg !17
  store %"Type<String>" %4, ptr @"Type<String>", align 8, !dbg !17
  br label %return, !dbg !17

return:                                           ; preds = %0
  ret void
}

define void @initialize.4() !dbg !18 {
  %1 = alloca %"Type<Type<Integer>>", align 8, !dbg !19
  %"Type<Type<Integer>>.name" = getelementptr inbounds %"Type<Type<Integer>>", ptr %1, i32 0, i32 0, !dbg !19
  %2 = call %String @string_from_c_string_and_length(ptr @4, i64 13), !dbg !20
  store %String %2, ptr %"Type<Type<Integer>>.name", align 8, !dbg !20
  %"Type<Type<Integer>>.size" = getelementptr inbounds %"Type<Type<Integer>>", ptr %1, i32 0, i32 1, !dbg !20
  %3 = call %Integer @integer_from_i64(i64 16), !dbg !20
  store %Integer %3, ptr %"Type<Type<Integer>>.size", align 8, !dbg !20
  %4 = load %"Type<Type<Integer>>", ptr %1, align 8, !dbg !20
  store %"Type<Type<Integer>>" %4, ptr @"Type<Type<Integer>>", align 8, !dbg !20
  br label %return, !dbg !20

return:                                           ; preds = %0
  ret void
}

define void @initialize.5() !dbg !21 {
  %1 = load %"Type<Integer>", ptr @"Type<Integer>", align 8, !dbg !22
  store %"Type<Integer>" %1, ptr @x, align 8, !dbg !22
  br label %return, !dbg !22

return:                                           ; preds = %0
  ret void
}

define void @initialize.6() !dbg !23 {
  %1 = load %Integer, ptr getelementptr inbounds (%"Type<Integer>", ptr @"Type<Integer>", i32 0, i32 1), align 8, !dbg !24
  %2 = call %Integer @clone_integer(%Integer %1), !dbg !24
  store %Integer %2, ptr @y, align 8, !dbg !24
  br label %return, !dbg !24

return:                                           ; preds = %0
  ret void
}

declare %Integer @clone_integer(%Integer)

define void @type_as_value.execute() !dbg !25 {
init_globals:
  call void @initialize(), !dbg !26
  call void @initialize.1(), !dbg !27
  call void @initialize.2(), !dbg !28
  call void @initialize.3(), !dbg !29
  call void @initialize.4(), !dbg !30
  call void @initialize.5(), !dbg !26
  call void @initialize.6(), !dbg !31
  br label %0, !dbg !31

0:                                                ; preds = %init_globals
  %1 = load %"Type<None>", ptr @"Type<None>", align 8, !dbg !27
  %2 = call %Integer @"size of <:Type<None>>"(%"Type<None>" %1), !dbg !27
  %3 = call %Integer @integer_from_i64(i64 0), !dbg !32
  %4 = call i1 @integer_eq_integer(%Integer %2, %Integer %3), !dbg !32
  call void @"println <:Bool>"(i1 %4), !dbg !32
  %5 = load %"Type<Bool>", ptr @"Type<Bool>", align 8, !dbg !28
  %6 = call %Integer @"size of <:Type<Bool>>"(%"Type<Bool>" %5), !dbg !28
  %7 = call %Integer @integer_from_i64(i64 1), !dbg !33
  %8 = call i1 @integer_eq_integer(%Integer %6, %Integer %7), !dbg !33
  call void @"println <:Bool>"(i1 %8), !dbg !33
  %9 = load %"Type<None>", ptr @"Type<None>", align 8, !dbg !34
  %10 = call %String @"<:Type<None>> as String"(%"Type<None>" %9), !dbg !34
  call void @"println <:String>"(%String %10), !dbg !34
  %11 = load %"Type<Type<Integer>>", ptr @"Type<Type<Integer>>", align 8, !dbg !30
  %12 = call %String @"<:Type<Type<Integer>>> as String"(%"Type<Type<Integer>>" %11), !dbg !30
  call void @"println <:String>"(%String %12), !dbg !30
  %13 = load %Integer, ptr @y, align 8, !dbg !35
  %14 = call %Integer @clone_integer(%Integer %13), !dbg !35
  %15 = load %"Type<Integer>", ptr @"Type<Integer>", align 8, !dbg !36
  %16 = call %Integer @"size of <:Type<Integer>>"(%"Type<Integer>" %15), !dbg !36
  %17 = call i1 @integer_eq_integer(%Integer %14, %Integer %16), !dbg !36
  call void @"println <:Bool>"(i1 %17), !dbg !36
  %18 = load %"Type<Integer>", ptr @"Type<Integer>", align 8, !dbg !37
  call void @"println <:Type<Integer>>"(%"Type<Integer>" %18), !dbg !37
  br label %return, !dbg !31

return:                                           ; preds = %0
  ret void
}

define void @"println <:Bool>"(i1 %0) !dbg !38 {
  %x = alloca i1, align 1
  store i1 %0, ptr %x, align 1
  %2 = load i1, ptr %x, align 1, !dbg !39
  %3 = call %String @"<:Bool> as String"(i1 %2), !dbg !39
  call void @"println <:String>"(%String %3), !dbg !39
  br label %return, !dbg !39

return:                                           ; preds = %1
  ret void
}

declare void @"println <:String>"(%String)

define %String @"<:Bool> as String"(i1 %0) !dbg !40 {
  %return_value = alloca %String, align 8
  %x = alloca i1, align 1
  store i1 %0, ptr %x, align 1
  br label %if.condition, !dbg !41

if.condition:                                     ; preds = %1
  %2 = load i1, ptr %x, align 1, !dbg !41
  br i1 %2, label %if.body, label %5, !dbg !41

if.body:                                          ; preds = %if.condition
  %3 = call %String @string_from_c_string_and_length(ptr @5, i64 4), !dbg !41
  store %String %3, ptr %return_value, align 8, !dbg !41
  br label %return, !dbg !41

return:                                           ; preds = %5, %if.body
  %4 = load %String, ptr %return_value, align 8
  ret %String %4

5:                                                ; preds = %if.condition
  %6 = call %String @string_from_c_string_and_length(ptr @6, i64 5), !dbg !41
  store %String %6, ptr %return_value, align 8, !dbg !41
  br label %return, !dbg !41
}

declare i1 @integer_eq_integer(%Integer, %Integer)

define %Integer @"size of <:Type<None>>"(%"Type<None>" %0) !dbg !42 {
  %return_value = alloca %Integer, align 8
  %ty = alloca %"Type<None>", align 8
  store %"Type<None>" %0, ptr %ty, align 8
  %size = getelementptr inbounds %"Type<None>", ptr %ty, i32 0, i32 1, !dbg !43
  %2 = load %Integer, ptr %size, align 8, !dbg !43
  store %Integer %2, ptr %return_value, align 8, !dbg !43
  br label %return, !dbg !43

return:                                           ; preds = %1
  %3 = load %Integer, ptr %return_value, align 8
  ret %Integer %3
}

define %Integer @"size of <:Type<Bool>>"(%"Type<Bool>" %0) !dbg !44 {
  %return_value = alloca %Integer, align 8
  %ty = alloca %"Type<Bool>", align 8
  store %"Type<Bool>" %0, ptr %ty, align 8
  %size = getelementptr inbounds %"Type<Bool>", ptr %ty, i32 0, i32 1, !dbg !45
  %2 = load %Integer, ptr %size, align 8, !dbg !45
  store %Integer %2, ptr %return_value, align 8, !dbg !45
  br label %return, !dbg !45

return:                                           ; preds = %1
  %3 = load %Integer, ptr %return_value, align 8
  ret %Integer %3
}

define %String @"<:Type<None>> as String"(%"Type<None>" %0) !dbg !46 {
  %return_value = alloca %String, align 8
  %ty = alloca %"Type<None>", align 8
  store %"Type<None>" %0, ptr %ty, align 8
  %name = getelementptr inbounds %"Type<None>", ptr %ty, i32 0, i32 0, !dbg !47
  %2 = load %String, ptr %name, align 8, !dbg !47
  store %String %2, ptr %return_value, align 8, !dbg !47
  br label %return, !dbg !47

return:                                           ; preds = %1
  %3 = load %String, ptr %return_value, align 8
  ret %String %3
}

define %String @"<:Type<Type<Integer>>> as String"(%"Type<Type<Integer>>" %0) !dbg !48 {
  %return_value = alloca %String, align 8
  %ty = alloca %"Type<Type<Integer>>", align 8
  store %"Type<Type<Integer>>" %0, ptr %ty, align 8
  %name = getelementptr inbounds %"Type<Type<Integer>>", ptr %ty, i32 0, i32 0, !dbg !49
  %2 = load %String, ptr %name, align 8, !dbg !49
  store %String %2, ptr %return_value, align 8, !dbg !49
  br label %return, !dbg !49

return:                                           ; preds = %1
  %3 = load %String, ptr %return_value, align 8
  ret %String %3
}

define %Integer @"size of <:Type<Integer>>"(%"Type<Integer>" %0) !dbg !50 {
  %return_value = alloca %Integer, align 8
  %ty = alloca %"Type<Integer>", align 8
  store %"Type<Integer>" %0, ptr %ty, align 8
  %size = getelementptr inbounds %"Type<Integer>", ptr %ty, i32 0, i32 1, !dbg !51
  %2 = load %Integer, ptr %size, align 8, !dbg !51
  store %Integer %2, ptr %return_value, align 8, !dbg !51
  br label %return, !dbg !51

return:                                           ; preds = %1
  %3 = load %Integer, ptr %return_value, align 8
  ret %Integer %3
}

define void @"println <:Type<Integer>>"(%"Type<Integer>" %0) !dbg !52 {
  %x = alloca %"Type<Integer>", align 8
  store %"Type<Integer>" %0, ptr %x, align 8
  %2 = load %"Type<Integer>", ptr %x, align 8, !dbg !53
  %3 = call %String @"<:Type<Integer>> as String"(%"Type<Integer>" %2), !dbg !53
  call void @"println <:String>"(%String %3), !dbg !53
  br label %return, !dbg !53

return:                                           ; preds = %1
  ret void
}

define %String @"<:Type<Integer>> as String"(%"Type<Integer>" %0) !dbg !54 {
  %return_value = alloca %String, align 8
  %ty = alloca %"Type<Integer>", align 8
  store %"Type<Integer>" %0, ptr %ty, align 8
  %name = getelementptr inbounds %"Type<Integer>", ptr %ty, i32 0, i32 0, !dbg !55
  %2 = load %String, ptr %name, align 8, !dbg !55
  store %String %2, ptr %return_value, align 8, !dbg !55
  br label %return, !dbg !55

return:                                           ; preds = %1
  %3 = load %String, ptr %return_value, align 8
  ret %String %3
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "ppl", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false, sysroot: "/")
!2 = !DIFile(filename: "type_as_value.ppl", directory: ".")
!3 = distinct !DISubprogram(name: "initialize", linkageName: "initialize", scope: !2, file: !2, line: 5, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!7 = !DILocation(line: 5, column: 8, scope: !3)
!8 = !DILocation(line: 0, scope: !3)
!9 = distinct !DISubprogram(name: "initialize.1", linkageName: "initialize.1", scope: !2, file: !2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!10 = !DILocation(line: 0, column: 17, scope: !9)
!11 = !DILocation(line: 0, scope: !9)
!12 = distinct !DISubprogram(name: "initialize.2", linkageName: "initialize.2", scope: !2, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!13 = !DILocation(line: 1, column: 17, scope: !12)
!14 = !DILocation(line: 0, scope: !12)
!15 = distinct !DISubprogram(name: "initialize.3", linkageName: "initialize.3", scope: !2, file: !2, line: 2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!16 = !DILocation(line: 2, column: 17, scope: !15)
!17 = !DILocation(line: 0, scope: !15)
!18 = distinct !DISubprogram(name: "initialize.4", linkageName: "initialize.4", scope: !2, file: !2, line: 3, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!19 = !DILocation(line: 3, column: 9, scope: !18)
!20 = !DILocation(line: 0, scope: !18)
!21 = distinct !DISubprogram(name: "initialize.5", linkageName: "initialize.5", scope: !2, file: !2, line: 5, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!22 = !DILocation(line: 5, column: 8, scope: !21)
!23 = distinct !DISubprogram(name: "initialize.6", linkageName: "initialize.6", scope: !2, file: !2, line: 6, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!24 = !DILocation(line: 6, column: 8, scope: !23)
!25 = distinct !DISubprogram(name: "type_as_value.execute", linkageName: "type_as_value.execute", scope: !2, file: !2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!26 = !DILocation(line: 5, column: 8, scope: !25)
!27 = !DILocation(line: 0, column: 17, scope: !25)
!28 = !DILocation(line: 1, column: 17, scope: !25)
!29 = !DILocation(line: 2, column: 17, scope: !25)
!30 = !DILocation(line: 3, column: 9, scope: !25)
!31 = !DILocation(line: 6, column: 8, scope: !25)
!32 = !DILocation(line: 0, column: 26, scope: !25)
!33 = !DILocation(line: 1, column: 26, scope: !25)
!34 = !DILocation(line: 2, column: 9, scope: !25)
!35 = !DILocation(line: 7, column: 8, scope: !25)
!36 = !DILocation(line: 7, column: 22, scope: !25)
!37 = !DILocation(line: 8, column: 8, scope: !25)
!38 = distinct !DISubprogram(name: "println <:Bool>", linkageName: "println <:Bool>", scope: !25, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!39 = !DILocation(line: 8, column: 15, scope: !38)
!40 = distinct !DISubprogram(name: "<:Bool> as String", linkageName: "<:Bool> as String", scope: !38, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!41 = !DILocation(line: 8, column: 15, scope: !40)
!42 = distinct !DISubprogram(name: "size of <:Type<None>>", linkageName: "size of <:Type<None>>", scope: !25, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!43 = !DILocation(line: 8, column: 15, scope: !42)
!44 = distinct !DISubprogram(name: "size of <:Type<Bool>>", linkageName: "size of <:Type<Bool>>", scope: !25, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!45 = !DILocation(line: 8, column: 15, scope: !44)
!46 = distinct !DISubprogram(name: "<:Type<None>> as String", linkageName: "<:Type<None>> as String", scope: !25, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!47 = !DILocation(line: 8, column: 15, scope: !46)
!48 = distinct !DISubprogram(name: "<:Type<Type<Integer>>> as String", linkageName: "<:Type<Type<Integer>>> as String", scope: !25, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!49 = !DILocation(line: 8, column: 15, scope: !48)
!50 = distinct !DISubprogram(name: "size of <:Type<Integer>>", linkageName: "size of <:Type<Integer>>", scope: !25, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!51 = !DILocation(line: 8, column: 15, scope: !50)
!52 = distinct !DISubprogram(name: "println <:Type<Integer>>", linkageName: "println <:Type<Integer>>", scope: !25, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!53 = !DILocation(line: 8, column: 15, scope: !52)
!54 = distinct !DISubprogram(name: "<:Type<Integer>> as String", linkageName: "<:Type<Integer>> as String", scope: !52, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!55 = !DILocation(line: 8, column: 15, scope: !54)
